File: src/ggrc_risk_assessments/assets/javascripts/models/risk_assessment.js
/*
 * Copyright (C) 2015 Google Inc., authors, and contributors <see AUTHORS file>
 * Licensed under http://www.apache.org/licenses/LICENSE-2.0 <see LICENSE file>
 * Created By: ivan@reciprocitylabs.com
 * Maintained By: ivan@reciprocitylabs.com
 */

/*
 */

(function (can) {

  _mustache_path = GGRC.mustache_path + "/risk_assessments";
  can.Model.Cacheable("CMS.Models.RiskAssessment", {
    root_object: "risk_assessment",
    root_collection: "risk_assessments",
    category: "risk_assessment",
    findAll: "GET /api/risk_assessments",
    findOne: "GET /api/risk_assessments/{id}",
    create: "POST /api/risk_assessments",
    update: "PUT /api/risk_assessments/{id}",
    destroy: "DELETE /api/risk_assessments/{id}",
    is_custom_attributable: true,
    attributes: {
      ra_manager: "CMS.Models.Person.stub",
      ra_counsel: "CMS.Models.Person.stub",
      context: "CMS.Models.Context.stub",
      documents: "CMS.Models.Document.stubs",
      program: "CMS.Models.Program.stub",
      modified_by: "CMS.Models.Person.stub",
      object_documents: "CMS.Models.ObjectDocument.stubs",
      custom_attribute_values : "CMS.Models.CustomAttributeValue.stubs",
    },
    tree_view_options: {
      attr_list: [
        {attr_title: 'Title', attr_name: 'title'},
        {attr_title: 'Code', attr_name: 'slug'},
        {attr_title: 'Effective Date', attr_name: 'start_date'},
        {attr_title: 'Stop Date', attr_name: 'end_date'},
        {
          attr_title: 'Risk Manager',
          attr_name: 'ra_manager',
          attr_sort_field: 'ra_manager.name|email'
        },
        {
          attr_title: 'Risk Counsel',
          attr_name: 'ra_counsel',
          attr_sort_field: 'ra_counsel.name|email'
        }
      ],
      add_item_view: _mustache_path + "/tree_add_item.mustache",
      child_options: [{
        model: "Document",
        mapping: "documents",
        show_view: _mustache_path + "/documents.mustache",
      }],
    },
    init: function () {
      this._super && this._super.apply(this, arguments);
      this.validateNonBlank("title");
      this.validateNonBlank("start_date");
      this.validateNonBlank("end_date");
    }
  }, {
    save: function () {
      // Make sure the context is always set to the parent program
      if (!this.context || !this.context.id) {
        this.attr('context', this.program.reify().context);
      }
      return this._super.apply(this, arguments);
    }
  });

})(window.can);

File: src/ggrc_risk_assessments/models/risk_assessment.py
# Copyright (C) 2013 Google Inc., authors, and contributors <see AUTHORS file>
# Licensed under http://www.apache.org/licenses/LICENSE-2.0 <see LICENSE file>
# Created By: dan@reciprocitylabs.com
# Maintained By: dan@reciprocitylabs.com

from ggrc import db
from ggrc.models.mixins import Base
from ggrc.models.mixins import CustomAttributable
from ggrc.models.mixins import Described
from ggrc.models.mixins import Noted
from ggrc.models.mixins import Slugged
from ggrc.models.mixins import Timeboxed
from ggrc.models.mixins import Titled
from ggrc.models.mixins import deferred
from ggrc.models.object_document import Documentable
from ggrc.models.person import Person
from ggrc.models.program import Program


class RiskAssessment(Documentable, Slugged, Timeboxed, Noted, Described,
                     CustomAttributable, Titled, Base, db.Model):
  __tablename__ = 'risk_assessments'
  _title_uniqueness = False

  ra_manager_id = deferred(
      db.Column(db.Integer, db.ForeignKey('people.id')), 'RiskAssessment')
  ra_manager = db.relationship(
      'Person', uselist=False, foreign_keys='RiskAssessment.ra_manager_id')

  ra_counsel_id = deferred(
      db.Column(db.Integer, db.ForeignKey('people.id')), 'RiskAssessment')
  ra_counsel = db.relationship(
      'Person', uselist=False, foreign_keys='RiskAssessment.ra_counsel_id')

  program_id = deferred(
      db.Column(db.Integer, db.ForeignKey('programs.id'), nullable=False),
      'RiskAssessment')
  program = db.relationship(
      'Program',
      backref='risk_assessments',
      uselist=False,
      foreign_keys='RiskAssessment.program_id')

  _fulltext_attrs = []

  _publish_attrs = [
      'ra_manager',
      'ra_counsel',
      'program',
  ]

  _aliases = {
      "ra_manager": {
        "display_name": "Risk Manager",
        "filter_by": "_filter_by_risk_manager",
      },
      "ra_counsel": {
        "display_name": "Risk Counsel",
        "filter_by": "_filter_by_risk_counsel",
      },
      "start_date": {
          "display_name": "Start Date",
          "mandatory": True,
      },
      "end_date": {
          "display_name": "End Date",
          "mandatory": True,
      },
      "program": {
          "display_name": "Program",
          "mandatory": True,
          "filter_by": "_filter_by_program",
      }
  }

  @classmethod
  def _filter_by_program(cls, predicate):
    return Program.query.filter(
        (Program.id == cls.program_id) &
        (predicate(Program.slug) | predicate(Program.title))
    ).exists()

  @classmethod
  def _filter_by_risk_manager(cls, predicate):
    return Person.query.filter(
      (Person.id == cls.ra_manager_id) &
      (predicate(Person.name) | predicate(Person.email))
    ).exists()

  @classmethod
  def _filter_by_risk_counsel(cls, predicate):
    return Person.query.filter(
      (Person.id == cls.ra_counsel_id) &
      (predicate(Person.name) | predicate(Person.email))
    ).exists()
